### PR TITLE
[RDY] Normalize result of out-of-bounds slicing

### PR DESCRIPTION
**Pull request by nhynes**
Note: this is a clean repush of #2808 that takes into account suggestions that deleting at an out-of-bounds index now returns an error. This is consistent with the behavior of plugin languages like Python.

Clamping the indices passed to the `_line_slice` commands to the start and end of the buffer using `normalize_index` leads to non-intuitive behavior. For instance, getting line(s) with a start index outside of the buffer's range returns the last line instead of an empty string or array, as is done by `getbufline`. Similarly, an out-of-bounds line/range set causes the last line to be clobbered instead of producing an error (c.f. `setline`).

This pull request addresses these issues by
- returning an empty string/array when the fetched start index is out of bounds,
- returning an error when setting/deleting an out-of-bounds location

### CODE REVIEW

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -179,6 +179,11 @@ void buffer_set_line_slice(Buffer buffer,
     return;
   }
 
+  if (!inbounds(buf, start)) {
+    api_set_error(err, Validation, _("Index out of bounds"));
+    return;
```
Comment (oni-link) [inline src/nvim/api/buffer.c]: Before this change `buffer_inser(buffer,-n-1,lines,err)` could be used to insert lines at the beginning of the buffer (n = number of lines in the buffer). After the change this is only possible with
`buffer_set_line_slice(buffer,0,0,true,false,replacement,err);`.

Comment (nhynes) [inline src/nvim/api/buffer.c]: Yes, but I daresay that the latter invocation is significantly less clear! Furthermore, since atomic atomic operations are currently unsupported, by the time that the plugin calculated `-n-1`, the actual line count could have changed. It seems dangerous to rely on this side effect.

File: src/nvim/api/buffer.c [hunk 2]
```diff
@@ -108,7 +108,7 @@ ArrayOf(String) buffer_get_line_slice(Buffer buffer,
   Array rv = ARRAY_DICT_INIT;
   buf_T *buf = find_buffer_by_handle(buffer, err);
 
-  if (!buf) {
+  if (!buf || !inbounds(buf, start)) {
```
Comment (tarruda) [inline src/nvim/api/buffer.c]: With this change, how does one gets the last N lines of a buffer without making two msgpack-rpc requests?

File: src/nvim/api/buffer.c [hunk 3]
```diff
@@ -179,6 +179,11 @@ void buffer_set_line_slice(Buffer buffer,
     return;
   }
 
+  if (!inbounds(buf, start)) {
```
Comment (tarruda) [inline src/nvim/api/buffer.c]: Same as above, how to make changes to the end of the buffer? For client-initiated "conversations", nvim doesn't block between requests, so one could potentially use an invalid line counts.

File: src/nvim/api/buffer.c [hunk 4]
```diff
@@ -550,3 +555,10 @@ static int64_t normalize_index(buf_T *buf, int64_t index)
   index = index > buf->b_ml.ml_line_count ? buf->b_ml.ml_line_count : index;
   return index;
 }
+
+// Determines whether a 0-based index is within the bounds of the buffer
+static Boolean inbounds(buf_T *buf, int64_t index)
```
Comment (justinmk) [inline src/nvim/api/buffer.c]: `bool`

Comment (nhynes) [inline src/nvim/api/buffer.c]: Good catch. I'll squash the commit before merge.

**Comment by justinmk** [comment]
@nhynes It should be fairly easy to write a test for this. It would also serve to precisely explain the expected behavior.

File: src/nvim/api/buffer.c (hunk 3 above)
Comment (splinterofchaos) [inline src/nvim/api/buffer.c]: Would it make sense to add empty lines up to `start`?

**Comment by nhynes** [comment]
@justinmk it has been done.

File: src/nvim/api/buffer.c (hunk 3 above)
Comment (nhynes) [inline src/nvim/api/buffer.c]: I'm not sure. While it would certainly be convenient, it's not consistent with vim commands like `setline` (if we're trying to emulate those) and it might be confusing for the behavior of `set_line_slice` to be entirely different depending on which end of the buffer has been overrun.

At least from my perspective, it would be much more common for an out-of-bounds set to result from plugin/user race conditions than from trying to set lines at a specific position in a file. I'd be interested in hearing of others' experiences, though.

It might just be easier and more explicit to just send over a chunk of vimscript/lua that performs the appropriate operations.

Comment (splinterofchaos) [inline src/nvim/api/buffer.c]: Yeah, I suppose I agree, though it occurs to me that the issue is that we don't have a way to append to a buffer. Perhaps this function should act like `setline`, but we should also have a `buffer_append` function to avoid race conditions.

Those that can be in a different PR.

Comment (nhynes) [inline src/nvim/api/buffer.c]: I think that you might already be able to accomplish this using `set_line_slice(buf, -1, -1, false, true, ["lines", "to", "append"])`

File: test/functional/api/buffer_spec.lua [hunk 5]
```diff
@@ -29,8 +29,14 @@ describe('buffer_* functions', function()
       eq('', curbuf('get_line', 0))
       curbuf('set_line', 0, 'line1')
       eq('line1', curbuf('get_line', 0))
+      eq('', curbuf('get_line', 1))
+      eq('', curbuf('get_line', -2))
+      eq(false, pcall(function() curbuf('set_line', 1) end))
+      eq(false, pcall(function() curbuf('set_line', -2) end))
```
Comment (justinmk) [inline test/functional/api/buffer_spec.lua]: This always returns false because set_line requires 3 arguments.

I'm merging this now, so I will fix it.

File: test/functional/api/buffer_spec.lua [hunk 6]
```diff
@@ -54,6 +60,15 @@ describe('buffer_* functions', function()
       eq({'a', 'b'}, curbuf('get_line_slice', 0, -1, true, false))
       eq({'b'}, curbuf('get_line_slice', 1, -1, true, false))
       eq({'b', 'c'}, curbuf('get_line_slice', -2, -1, true, true))
+      eq({}, curbuf('get_line_slice', 2, 3, false, true))
```
Comment (justinmk) [inline test/functional/api/buffer_spec.lua]: I noticed that over-shooting the buffer contents will "wrap around" an unlimited number of times. This is a bit surprising. For example, in this test, the buffer contents are:

```
a
b
c
```

Then, `get_line_slice(1,18, true, true)` returns `['b', 'c']`. @nhynes is this how python works as well?

Comment (justinmk) [inline test/functional/api/buffer_spec.lua]: I also noticed that 

```
get_line_slice(0, 0, false, false)
get_line_slice(0, 0, true, false)
get_line_slice(0, 0, false, true)

return empty array, yet get_line_slice(0, 0, true, true) returns `['a']`. This makes me wonder why we have the `include_start` and `include_end` parameters. Are they really necessary?

**Comment by justinmk** [comment]
Merged with some edits, and open questions.

File: test/functional/api/buffer_spec.lua (hunk 6 above)
Comment (nhynes) [inline test/functional/api/buffer_spec.lua]: @justinmk
Re: your first question. This is, indeed, how python works as well:
```python
lines = ['a', 'b', 'c']
lines[1:99] == ['b', 'c']
```
This behavior is also present in several other dynamic languages like Perl, JavaScript, and Ruby, so it's not without precedent. While, the compiled languages that I tried (Go, Java/Scala, Haskell, Octave, and Julia), all throw errors when slicing using an out-of-bounds index, it's not possible to mimic this behavior without transactions.

Re: `include_(start|end)`, I'm in favor of removing them since it's not clear what "start" and "end" mean. For instance, with the lines `[a, b, c]`, should 
`get_line_slice(1, 99, true, false)` return `[b, c]` or just `[b]`?
I think that having the client explicitly make these decisions would improve clarity.
